processing/src/main/java/org/apache/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -314,7 +315,12 @@ protected IncrementalIndex(
     for (DimensionSchema dimSchema : dimensionsSpec.getDimensions()) {
       ValueType type = TYPE_MAP.get(dimSchema.getValueType());
       String dimName = dimSchema.getName();
-      ColumnCapabilitiesImpl capabilities = makeDefaultCapabilitiesFromValueType(type);
+
+      // Note: Things might be simpler if DimensionSchema had a method "getColumnCapabilities()" which could return
+      // type specific capabilities by itself. However, for various reasons, DimensionSchema currently lives in druid-core
+      // while ColumnCapabilities lives in druid-processing which makes that approach difficult.
+      ColumnCapabilitiesImpl capabilities = makeDefaultCapabilitiesFromValueType(type, dimSchema.getTypeName());
```

Comment:
Apologies for the conflicts here, I made some semi disruptive changes to push `ColumnCapabilities` into the `DimensionIndexer` implementations so they can be more accurate. I think the changes should still be workable with your addition of `DimensionHandlerProvider`, just the dimension indexer it provides will need to provide the complex column capabilities.

Reply:
it should be alright, I will update this PR later when I have my MapStringString extension ready/deployed and tested with changes in this patch. thanks for the heads up.